### DISCUSSION TITLE
【llbc】对象池改进：增加对象释放的便利性

### DISCUSSION THREAD

**Discussion by lailongwei** [lailongwei/llbc#302] [General]
当前对象池Acquire出来的对象，如果想通过`LLBC_Recycle()`进行释放，有两种途径：
* 继承`LLBC_PoolObject`：将自动拥有**记录来自哪个对象池**的能力
* 实现`python like protocol`的两个方法：`GetTypedObjPool()`、`SetTypedObjPool()`，**让对象能记录自己来自哪个对象池**

但：对于stl容器、外部未适配上面两种机制的类，将无法享受到`LLBC_Recycle()`的便利性，甚至**一旦使用将引发crash**，基于这个情况，**对对象池性能作让步，以 提升通用性 及 减少心智负担**：
* `Acquire()`方法返回调整为`std::shared_ptr`，以统一管理内存的释放（可使用recycle，但具体LLBC_Recycle如何实现需要再细化）
* 提供`RawAcquire()`方法（或有更好的方法命名可以探讨），以支持高效使用

**Comment by barton-git** [comment]
是否考虑增加AcquireUnique()方法返回unique_ptr，完整的支持所有智能指针

**Comment by lailongwei** [comment]
> 是否考虑增加AcquireUnique()方法返回unique_ptr，完整的支持所有智能指针

这个问题，我先转成`discussion`

**Comment by lailongwei** [comment]
如题，这边的想法是：
* 确保接口的简介性，即想永远只保留一个方法：`Acquire()`
* 做到：**裸返回**、**shared_ptr包装返回**、**unique_ptr包装返回**上，想通过模板参数来达到，具体：
```c++
template <typename Obj, typename RetType = Obj *>
RetType Acquire();

// 在RetType = Obj *的时候，就是裸返回，适配`LLBC_Object子类`、`实现对象池可重用协议的两个方法`
// RetType = shared_ptr/unique_ptr的时候，就是包装返回
```

调用的时候，RetType的默认值可以是`std::shared_ptr<Obj>`，默认就是shared_ptr，其它情况显式指定，但缺乏灵活性，可能还需要进一步改进

@barton-git

**Reply by barton-git** [reply]
通过模版参数区分的方式可行，业务也能感知，不过模版参数RetType换成enum类型: kShared, kRaw, kUnique会更直观，使用也简单点

**Reply by lailongwei** [reply]
> 通过模版参数区分的方式可行，业务也能感知，不过模版参数RetType换成enum类型: kShared, kRaw, kUnique会更直观，使用也简单点

确实如此，不过具体可行性是否能做到，还需要实际验证一下，因为ret不是函数签名区分，如果 @barton-git 有空，可以帮忙改进一下，这个应该不急

**Reply by barton-git** [reply]
通过增加一个类型帮助的struct应该可以做到，下面是初步设想的伪代码


enum class PtrType{
    kRaw,
    kShared, 
    kUnique
};

template<typename Obj,  PtrType>
struct PtrWrraper
{
};

// 偏特化决定ptr类型
template< typename Obj >
struct PtrWrraper<Obj, kShared>
{
    using WrraperType = std::shared_ptr<Obj>;

    static WrraperType WrrapPtr(Obj* p)
    {
        // 使用shared_ptr返回
        return std::shared_ptr<Obj>(p, deleter);
    }
}

// unique_ptr
// raw_ptr

// 外侧acquire定义
template<typename Obj,  PtrType>
PtrWrraper<Obj, PtrType>::WrraperType
Acquire()
{
    // get obj-ptr from pool
    return PtrWrraper<Obj, PtrType>::WrrapPtr(obj_ptr);
}

**Comment by lailongwei** [comment]
针对上面的方案，有没有可能去除伪代码中的`PtrWrapper`？想直接得到就是类型，而不需要多一层包装，想**用将最NB的技术跟方案，以最原始、简单的方式呈现给开发者**

**Reply by barton-git** [reply]
暂时还想不到，除非Acquire的模版参数中指定返回类型
